Ecotone/Dencun implementation changes

EvanJRichard opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

This implements the changes described in #8701 

Fixes https://github.com/ethereum-optimism/protocol-quest/issues/68
Fixes https://github.com/ethereum-optimism/protocol-quest/issues/69

Commit:
op-node: L2 Dencun implementation updates

init branch

Update reference to Eclipse/Ecotone in specs.

Pull in Danyal's deposit source and add additional tests.

Add notion of ParentBeaconRoot and build a contract deploy tx in attributes.go.

Add a test for activating dencun l2 fork after genesis.

Add draft ecotone setup.

Add first pass of Eclipse upgrade txns

Fix tests/compile

Code review feedback

Obey linter.

Move ecotone setup to helpers.go; get the ParentBeaconBlockRoot from the l1Info in attributes.go.

chore(op-node): Add tests for Ecotone deposit transactions (#8746)

* Source hash teEvanJRichard <evan@oplabs.co>

dencun review fixes

derive: ecotone upgradeTo abi encoding

op-e2e: test L2 exclusion of blob-txs in Ecotone

op-node/rollup: deduplicate ecotone activation helper func, fix rollup config var name

op-chain-ops: clarify 4788 contract nonce

op-node/rollup: add setEcotone to ecotone upgrade txs

dencun review fixes

Dencun: P2P / EngineAPI / ExecutionPayloadEnvelope changes

Includes:
- Pass through execution payload (Envelope type everywhere) by Danyal,
  extended by Proto
- Fix ecotone upgrade txns, by Danyal
- ci fixes by Danyal
- P2P Req/Resp (version based encoding/decoding) by Danyal
- EngineAPI v3 usage by Danyl, rebased by Proto on EngineController (from
  trianglesphere)
- Block v3 Gossip validation, by Danyal
- Block v3 Gossip publishing, by Proto

Rebased on updated Ecotone / Dencun base branch

op-e2e: fix upgrade-txs count in test

op-node: fix l1 info scalar migration, implement dencun review suggestions

op-node: more dencun review nit fixes

op-node: rabbit suggestions, but fixed

Fix nil pointer in p2p sync for Ecotone blocks

dencun: fix more nits

op-e2e: fix lint

Co-authored-by: Danyal Prout <me@dany.al>
Co-authored-by: protolambda <proto@protolambda.com>
Co-authored-by: EvanJRichard <evan@oplabs.co>

EvanJRichard commented on op-chain-ops/genesis/layer_two.go:
```diff
@@ -75,5 +75,14 @@ func BuildL2Genesis(config *DeployConfig, l1StartBlock *types.Block) (*core.Gene
 	}
 
-	return db.Genesis(), nil
+	genesis := db.Genesis()
+
+	if config.EcotoneTime(genesis.Timestamp) == &genesis.Timestamp {
+		err = setupEcotone(db)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return genesis, nil
```
@trianglesphere This came to mind during tonight's office hours discussion "how do we handle deployment of forks-at-genesis that involve contracts", is it relevant?

EvanJRichard commented:
Feedback from #8746: should use `op-node/rollup/derive:EcotoneNetworkUpgradeTransactions()` inside genesis generation

EvanJRichard commented on op-node/rollup/derive/attributes.go:
```diff
@@ -100,14 +99,23 @@ func (ba *FetchingAttributesBuilder) PreparePayloadAttributes(ctx context.Contex
+	}
+
 	l1InfoTx, err := L1InfoDepositBytes(seqNumber, l1Info, sysConfig, ba.cfg.IsRegolith(nextL2Time))
 	if err != nil {
 		return nil, NewCriticalError(fmt.Errorf("failed to create l1InfoTx: %w", err))
 	}
 
-	txs := make([]hexutil.Bytes, 0, 1+len(depositTxs))
+	txs := make([]hexutil.Bytes, 0, 1+len(depositTxs)+len(upgradeTxs))
 	txs = append(txs, l1InfoTx)
 	txs = append(txs, depositTxs...)
+	txs = append(txs, upgradeTxs...)
```
TODO: consider putting `upgradeTxs` first!

tynes replied:
The spec says it is correct as is: https://github.com/ethereum-optimism/optimism/pull/8602/files#diff-882757e7a296e749f616fb3e5972b2f8041dc9c515876529981fddd38e73a6a7R1064

EvanJRichard commented on op-node/rollup/driver/sequencer.go:
```diff
@@ -91,6 +91,11 @@ func (d *Sequencer) StartBuildingBlock(ctx context.Context) error {
 
+	// For the Ecotone activation block we shouldn't include any sequencer transactions.
+	if d.config.IsEcotoneUpgradeDepositBlock(l2Head.Time, uint64(attrs.Timestamp)) {
+		attrs.NoTxPool = true
```
log this `info`

protolambda replied:
added

sebastianst commented on op-node/rollup/derive/attributes.go:
```diff
@@ -122,5 +130,6 @@ func (ba *FetchingAttributesBuilder) PreparePayloadAttributes(ctx context.Contex
 		NoTxPool:              true,
 		GasLimit:              (*eth.Uint64Quantity)(&sysConfig.GasLimit),
```
Do the `EcotoneNetworkUpgradeTransactions` consume a deterministic amount of gas, or is there at least a deterministic upper bound? If yes, we could add this amount to the `GasLimit` to ensure user deposit transactions have the same amount of gas available in the upgrade block.

We could let `EcotoneNetworkUpgradeTransactions` return this gas limit above and do something like

```go
gaslimit := (eth.Uint64Quantity)(sysConfig.GasLimit)

    // on line 104
    upgradeTxs, upgradeGas, err = EcotoneNetworkUpgradeTransactions()
    gaslimit += upgradeGas
```
before and then here
```suggestion
		GasLimit:              &gaslimit,
```
So in this sense, we would build up the `gaslimit` value just as we build up the `txs`.

tynes replied:
This is a good idea, I believe that this needs to be specified better. They should consume a deterministic amount of gas cc @danyalprout 

The spec says TBC for the gas values of the individual txs, this is under specified still

tynes replied:
Actually I don't think we want to modify the gas limit here, it may break things. Not sure how this interacts with 1559. Thoughts @protolambda @trianglesphere @ajsutton ?

protolambda replied:
I'll add a todo issue to review the gas-limits of the other upgrade txs. The gas-limit of the only L2-Dencun-specific tx is a fixed constant, defined in the EIP, and very low.

protolambda replied:
Added to https://github.com/ethereum-optimism/protocol-quest/issues/62 description

tynes commented on op-chain-ops/genesis/setters.go:
```diff
@@ -89,3 +89,12 @@ func setupPredeploy(db vm.StateDB, deployResults immutables.DeploymentResults, s
 
 	return nil
 }
+
```
Could we add a better comment here? This is the contract that holds the ring buffer of CL roots right? Perhaps the source of truth for that code should be superchain registry?

protolambda replied:
added, it's a very short contract, defined by an EIP, rather than solidity source-code. I think it can just be a constant.

tynes commented on op-node/rollup/derive/deposit_source_test.go:
```diff
@@ -0,0 +1,70 @@
+# source hash type:
+# 0x0000000000000000000000000000000000000000000000000000000000000002
+# compute source hash:
+cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002ab0dfc96b47739a0ae1d415bbfaae79ebb1111861a3b7cfbbaa6ca4a9e618357
```
Perhaps link to spec as it contains a 1 liner that computes the hash using `cast concat-hex` and subshells or copy the 1 liner here for simplicity in verification

https://github.com/ethereum-optimism/optimism/pull/8602/files

tynes commented on op-node/rollup/derive/ecotone_upgrade_transactions.go:
```diff
@@ -0,0 +1,132 @@
+		return nil, err
+	}
+
+	upgradeL1BlockData, err := proxyAbi.Pack("upgradeTo", newL1BlockAddress)
```
Thoughts on removing the abi package from consensus? We have this package used elsewhere for abi encoding in consensus https://github.com/ethereum-optimism/optimism/blob/develop/op-service/solabi/util.go

Since we want to remove the bindings longer term, they are really heavy for consensus

EvanJRichard replied:
I like trying to keep things "light", but also like how this block of code reads using the proxy ABI bindings. I think it's worth it to keep this in for now, and then move to `op-service:solabi` as part of removing the bindings.

protolambda replied:
updated it to use `solabi`, I don't think we should depend on op-bindings directly, since we are hoping to remove those, and the ABI call is really simple here.

tynes commented on op-node/rollup/derive/ecotone_upgrade_transactions.go:
```diff
@@ -0,0 +1,132 @@
+		From:                common.Address{},
+		To:                  &predeploys.L1BlockAddr,
+		Value:               nil,
+		Gas:                 200_000,
```
The gas values are underspecified still

tynes replied:
https://github.com/ethereum-optimism/client-pod/issues/421

protolambda replied:
The non-dencun gas amounts are tracked by the issue, the gas of the dencun 4788 tx is a specific value as defined in the EIP. Resolving this, for the purposes of this PR.

tynes commented on op-node/rollup/derive/deposit_source.go:
```diff
@@ -44,3 +45,16 @@ func (dep *L1InfoDepositSource) SourceHash() common.Hash {
 	copy(domainInput[32:], depositIDHash[:])
 	return crypto.Keccak256Hash(domainInput[:])
 }
+
```
Would love godocs here

protolambda replied:
added

tynes commented on op-node/rollup/derive/deposit_source_test.go:
```diff
@@ -0,0 +1,70 @@
+	assert.Equal(t, expected, actual.Hex())
+}
+
+/*
```
nit: this doesn't follow standard godoc style

tynes commented on op-node/rollup/derive/ecotone_upgrade_transactions.go:
```diff
@@ -0,0 +1,132 @@
+
+	deployEIP4788, err := types.NewTx(&types.DepositTx{
+		From: eip4788From,
+		// to is null
```
thoughts on just setting it to nil in code?

tynes commented on op-node/rollup/derive/ecotone_upgrade_transactions.go:
```diff
@@ -0,0 +1,132 @@
+		// to is null
+		Mint:                big.NewInt(0),
+		Value:               big.NewInt(0),
+		Gas:                 0x3d090,
```
This is in hex and the others are in decimal

protolambda replied:
This is because the EIP-4788 specifies it in hexadecimal, and so it's easier to cross-verify. We should add a comment mentioning the EIP definition of this transaction.

protolambda commented:
rebased on develop (and pulling in latest op-geth and superchain registry)

protolambda commented:
Rebased on latest `develop` again

protolambda commented:
Almost ready, just need the https://github.com/ethereum-optimism/optimism/pull/8898 changes, and then a rebase on `develop` to incorporate the engine-controller change.

trianglesphere commented on op-chain-ops/genesis/setters.go:
```diff
@@ -89,3 +89,20 @@ func setupPredeploy(db vm.StateDB, deployResults immutables.DeploymentResults, s
+	eip4788ContractCode := common.Hex2Bytes("0x3373fffffffffffffffffffffffffffffffffffffffe14604d57602036146024575f5ffd5b5f35801560495762001fff810690815414603c575f5ffd5b62001fff01545f5260205ff35b5f5ffd5b62001fff42064281555f359062001fff015500")
+	db.CreateAccount(eip4788ContractAddr)
+	db.SetCode(eip4788ContractAddr, eip4788ContractCode)
+	db.SetNonce(eip4788ContractAddr, 1) // TODO review why goerli nonce is 1
```
Was this already deployed on goerli? This feels like a pretty important todo to figure out.

tynes replied:
This is expected behavior for it to be 1, because you can deploy no code it increments the nonce of the account. see https://github.com/ethereum/go-ethereum/blob/daa2e5d6a66833b9834b60a3a46835610bbde99a/core/vm/evm.go#L450

trianglesphere commented on op-node/rollup/derive/ecotone_upgrade_transactions_test.go:
```diff
@@ -0,0 +1,110 @@
+			expectedHash: "0x69b763c48478b9dc2f65ada09b3d92133ec592ea715ec65ad6e7f3dc519dc00c",
+		},
+	} {
+		assert.Equal(t, common.HexToHash(test.expectedHash), test.source.SourceHash())
```
nit: We tend to only use the `require` package

trianglesphere commented on op-e2e/actions/dencun_fork_test.go:
```diff
@@ -97,3 +102,140 @@ func TestDencunL1ForkAtGenesis(gt *testing.T) {
+	require.True(t, sd.L2Cfg.Config.IsCancun(l2Head.Number, l2Head.Time), "Cancun active")
+	require.NotNil(t, l2Head.ExcessBlobGas, "Cancun blob gas in header")
+
+	// try to build an L2 block with a blob tx, the EVM should fail
```
Delete this comment? It appears we already test this?

trianglesphere commented on op-node/rollup/derive/ecotone_upgrade_transactions.go:
```diff
@@ -0,0 +1,136 @@
+		Value:               nil,
+		Gas:                 300_000,
+		IsSystemTransaction: false,
+		Data:                common.FromHex(bindings.L1BlockMetaData.Bin),
```
Is this metadata stable enough?

tynes replied:
this is not stable enough, we should copy in a string. we should define the string as a `const` and in the godoc show how to reproduce the string. something like: "commit ..., `jq .bytecode < forge-artifacts/...`"

trianglesphere reviewed (commented):
Looking pretty good. This does seem like it may have some conflicts with https://github.com/ethereum-optimism/optimism/pull/8786

tynes commented on op-e2e/actions/ecotone_fork_test.go:
```diff
@@ -0,0 +1,105 @@
+	contractAddy := common.HexToAddress("0xA83144c24063155FBcBF8E51f96dbD4378e18907")
+	code, err := engine.EthClient().CodeAt(context.TODO(), contractAddy, latestBlock.Number())
+	require.NoError(t, err)
+	fmt.Println("code", code)
```
why this log?

protolambda replied:
will clean up the upgrade txs

tynes commented on op-node/rollup/derive/attributes.go:
```diff
@@ -100,14 +99,23 @@ func (ba *FetchingAttributesBuilder) PreparePayloadAttributes(ctx context.Contex
 
+	var upgradeTxs []hexutil.Bytes
+	if ba.cfg.IsEcotoneUpgradeDepositBlock(l2Parent.Time, nextL2Time) {
+		upgradeTxs, err = EcotoneNetworkUpgradeTransactions()
```
This function is going to need to have the system config passed in so that it can migrate the scalar to the basefeeScalar

protolambda replied:
Roberto implemented that base fee scalar update below

protolambda commented:
Rebased on develop, resolved a few merge conflicts with other ecotone work

protolambda commented on op-e2e/system_test.go:
```diff
@@ -1198,7 +1198,7 @@ func TestFees(t *testing.T) {
 	require.Nil(t, err)
 
-	l1Fee := l1CostFn(receipt.BlockNumber.Uint64(), header.Time, tx.RollupDataGas(), tx.IsSystemTx())
+	l1Fee := l1CostFn(tx.RollupCostData(), header.Time)
```
bad rabbit, that is a line number, not Go code

tynes commented on op-chain-ops/genesis/layer_two.go:
```diff
@@ -112,5 +112,10 @@ func BuildL2Genesis(config *DeployConfig, l1StartBlock *types.Block) (*core.Gene
 		}
 	}
 
```
I would much prefer to have a top level function `setupNetworkUpgrades` or `setupHardforks` and then we move the branching logic that is inside of `setupEcotone` outside of `setupEcotone` into `setupNetworkUpgrades`

tynes commented on op-e2e/actions/ecotone_fork_test.go:
```diff
@@ -0,0 +1,105 @@
+		require.Equal(t, types.ReceiptStatusSuccessful, receipt.Status)
+	}
+
+	l1BlockDeployerAddress := common.HexToAddress("0x4210000000000000000000000000000000000000")
```
these should be defined in a place where there is a single source of truth

tynes commented on op-e2e/actions/ecotone_fork_test.go:
```diff
@@ -0,0 +1,105 @@
+	assert.Greater(t, cost.Uint64(), uint64(0))
+
+	// 4788 contract is deployed
+	code, err := engine.EthClient().CodeAt(context.Background(), common.HexToAddress("0x000F3df6D732807Ef1319fB7B8bB8522d0Beac02"), latestBlock.Number())
```
this address should use single source of truth

tynes commented on op-node/rollup/derive/ecotone_upgrade_transactions.go:
```diff
@@ -0,0 +1,152 @@
+		From:                L1InfoDepositerAddress,
+		To:                  &predeploys.GasPriceOracleAddr,
+		Value:               nil,
+		Gas:                 1_000_000, // TBD
```
some values have `nil` some have `big.NewInt(0)` - should be standard

protolambda replied:
fixed

tynes commented on op-bindings/predeploys/addresses_test.go:
```diff
@@ -51,6 +51,7 @@ func TestL1BlockSlots(t *testing.T) {
 	require.Equal(t, types.OverheadSlot, overHeadSlot)
 	require.Equal(t, types.ScalarSlot, scalarSlot)
 	require.Equal(t, types.L1BasefeeSlot, l1BaseFeeSlot)
+
```
nit: useless diff

tynes commented on op-chain-ops/genesis/helpers.go:
```diff
@@ -107,3 +108,18 @@ func newHexBig(in uint64) *hexutil.Big {
 	return &hb
 }
+
+func setupEcotone(config *DeployConfig, db *state.MemoryStateDB) (*state.MemoryStateDB, error) {
```
The implementation describes a function called `maybeSetupEcotone` since its conditional

tynes commented on op-e2e/actions/ecotone_fork_test.go:
```diff
@@ -0,0 +1,105 @@
+	for i := 1; i < 5; i++ {
+		txn := transactions[i]
+		receipt, err := engine.EthClient().TransactionReceipt(context.Background(), txn.Hash())
+		fmt.Println("transaction", i, receipt.GasUsed)
```
nit: no printing leftover

tynes commented:
Mostly nits but the big one is finalization of the upgrade txs. The final calldata and gaslimits must be declared in the specs as well as standardizing on `nil` vs non `nil` values for values that are not necessary (ie mint)

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -24,6 +24,10 @@ func (_m *Consensus) EXPECT() *Consensus_Expecter {
 func (_m *Consensus) AddNonVoter(id string, addr string) error {
 	ret := _m.Called(id, addr)
 
+	if len(ret) == 0 {
+		panic("no return value specified for AddNonVoter")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -67,6 +71,10 @@
 func (_m *Consensus) AddVoter(id string, addr string) error {
 	ret := _m.Called(id, addr)
 
+	if len(ret) == 0 {
+		panic("no return value specified for AddVoter")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -110,6 +118,10 @@
 func (_m *Consensus) CommitUnsafePayload(payload *eth.ExecutionPayload) error {
 	ret := _m.Called(payload)
 
+	if len(ret) == 0 {
+		panic("no return value specified for CommitUnsafePayload")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -152,6 +164,10 @@
 func (_m *Consensus) DemoteVoter(id string) error {
 	ret := _m.Called(id)
 
+	if len(ret) == 0 {
+		panic("no return value specified for DemoteVoter")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -194,6 +210,10 @@
 func (_m *Consensus) LatestUnsafePayload() *eth.ExecutionPayload {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for LatestUnsafePayload")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -237,6 +257,10 @@
 func (_m *Consensus) Leader() bool {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for Leader")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -278,6 +302,10 @@
 func (_m *Consensus) LeaderCh() <-chan bool {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for LeaderCh")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -321,6 +349,10 @@
 func (_m *Consensus) LeaderWithID() (string, string) {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for LeaderWithID")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -372,6 +404,10 @@
 func (_m *Consensus) RemoveServer(id string) error {
 	ret := _m.Called(id)
 
+	if len(ret) == 0 {
+		panic("no return value specified for RemoveServer")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -414,6 +450,10 @@
 func (_m *Consensus) ServerID() string {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for ServerID")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -455,6 +495,10 @@
 func (_m *Consensus) Shutdown() error {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for Shutdown")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -496,6 +540,10 @@
 func (_m *Consensus) TransferLeader() error {
 	ret := _m.Called()
 
+	if len(ret) == 0 {
+		panic("no return value specified for TransferLeader")
+	}
```
bad rabbit, this is generated code for test mocks

protolambda commented on op-conductor/consensus/mocks/Consensus.go:
```diff
@@ -537,6 +585,10 @@
 func (_m *Consensus) TransferLeaderTo(id string, addr string) error {
 	ret := _m.Called(id, addr)
 
+	if len(ret) == 0 {
+		panic("no return value specified for TransferLeaderTo")
+	}
```
bad rabbit, this is generated code for test mocks

danyalprout committed:
Add tests for attribute matching

protolambda commented:
rebased on #8981

danyalprout committed:
Provide a no-op blob fetcher to prevent derivation error

Fail tests when there is an unknown error

Fix typo / empty array to nil

Update L2 tests to ecotone / add additional checks

axelKingsley commented on op-conductor/conductor/service.go:
```diff
@@ -559,7 +559,9 @@ func (oc *OpConductor) startSequencer() error {
 		if uint64(unsafeInCons.BlockNumber)-unsafeInNode.NumberU64() == 1 {
 			// tries to post the unsafe head to op-node when head is only 1 block behind (most likely due to gossip delay)
-			if err = oc.ctrl.PostUnsafePayload(context.Background(), unsafeInCons); err != nil {
+			// TODO:
```
?

protolambda replied:
related to the below breaking change comment to `PostUnsafePayload`. Will update the comment to refer to a ticket. Was unsure whether or not we were going to address it as part of this PR. But since that thing isn't in prod yet, we shouldn't block.

trianglesphere replied:
I'd expect this to not pass CI b/c of the dangling todo

protolambda commented on op-e2e/system_test.go:
```diff
@@ -188,7 +188,7 @@ func TestSystemE2EDencunAtGenesisWithBlobs(t *testing.T) {
 	// send a blob-containing txn on l1
 	ethPrivKey := sys.Cfg.Secrets.Alice
-	txData := transactions.CreateEmptyBlobTx(ethPrivKey, true, sys.Cfg.L1ChainIDBig().Uint64())
+	txData := transactions.CreateEmptyBlobTx(true, sys.Cfg.L1ChainIDBig().Uint64())
```
it's used right below in `MustSignNewTx`

protolambda committed:
op-e2e: dencun action-test setup fixes

danyalprout commented on op-node/node/api.go:
```diff
@@ -71,16 +71,17 @@ func (n *adminAPI) SequencerActive(ctx context.Context) (bool, error) {
-func (n *adminAPI) PostUnsafePayload(ctx context.Context, payload *eth.ExecutionPayload) error {
+func (n *adminAPI) PostUnsafePayload(ctx context.Context, envelope *eth.ExecutionPayloadEnvelope) error {
 	recordDur := n.M.RecordRPCServerRequest("admin_postUnsafePayload")
 	defer recordDur()
 
-	if actual, ok := payload.CheckBlockHash(); !ok {
+	payload := envelope.ExecutionPayload
+	if actual, ok := envelope.CheckBlockHash(); !ok {
 		log.Error("payload has bad block hash", "bad_hash", payload.BlockHash.String(), "actual", actual.String())
 		return fmt.Errorf("payload has bad block hash: %s, actual block hash is: %s", payload.BlockHash.String(), actual.String())
 	}
 
-	return n.dr.OnUnsafeL2Payload(ctx, payload)
+	return n.dr.OnUnsafeL2Payload(ctx, envelope)
```
Do you think we should make a V2 API for this instead of making this breaking change 

I'm leaning towards making the breaking change as it was added [recently](https://github.com/ethereum-optimism/optimism/pull/8800) and for op-ha which isn't used yet.

protolambda replied:
If it's not used yet we should just make the breaking change. The envelope type will allow us to avoid breaking changes in the future.

trianglesphere replied:
@0x00101010 can you confirm? I also think we can likely perform the breaking change here.

0x00101010 replied:
Yes, it's fine. We can rework after this change is in. It'll be better to avoid future changes 👍

roberto-bayardo reviewed (approved):
This LGTM though I don't have as much context on this as some of the other PRs.

trianglesphere commented on op-node/p2p/sync.go:
```diff
@@ -608,8 +610,8 @@ func (s *SyncClient) doRequest(ctx context.Context, id peer.ID, expectedBlockNum
 	version := binary.LittleEndian.Uint32(versionData[:])
-	if version != 0 {
-		return fmt.Errorf("unrecognized ExecutionPayload version: %d", version)
+	if version != 0 && version != 1 {
```
Should this be an OR?

ajsutton replied:
Wouldn't an OR  always evaluate to true - if version is 0 it's != 1 and if it's 1 it's != 0 so one branch would always be true.  I am notoriously bad at boolean logic though...

trianglesphere commented on op-node/rollup/derive/engine_controller.go:
```diff
@@ -307,6 +309,6 @@ func (e *EngineController) ForkchoiceUpdate(ctx context.Context, state *eth.Fork
 // NewPayload implements LocalEngineControl.
-func (e *EngineController) NewPayload(ctx context.Context, payload *eth.ExecutionPayload) (*eth.PayloadStatusV1, error) {
-	return e.engine.NewPayload(ctx, payload)
+func (e *EngineController) NewPayload(ctx context.Context, payload *eth.ExecutionPayload, parentBeaconBlockRoot *common.Hash) (*eth.PayloadStatusV1, error) {
```
Why doesn't new payload take the envelope?

protolambda replied:
Because the Engine API RPC definition of NewPayload does it this way

trianglesphere reviewed (commented):
This broadly looks good (though there are a couple small things I found). This was pretty hard to review b/c it had so many changes over a large number of files.

ajsutton commented on op-node/p2p/gossip.go:
```diff
@@ -309,26 +325,58 @@ func BuildBlocksValidator(log log.Logger, cfg *rollup.Config, runCfg GossipRunti
+			}
+		}
+
+		// [REJECT] if the block is on a topic >= V3 and the parent beacon block root is nil
+		if blockVersion == eth.BlockV3 && envelope.ParentBeaconBlockRoot == nil {
```
Slightly suspect that the comment says `>= V3` and the code says `== V3`. Maybe we should have a `HasParentBeaconBlockRoot` method on the block version like we do for the other blobs and withdrawals?

protolambda committed:
op-node: op-conductor dencun todo

protolambda commented:
Rebased on ecotone-upgrade PR, fixed that op-conductor related TODO comment to reference an issue

protolambda committed:
dencun: fix review nit about parent beacon block root gossip check style

protolambda commented:
rebased on develop, base ecotone-upgrade PR was merged
